dev/buildtool/inspection_commands.py

```diff
@@ -12,100 +12,1154 @@
+    self.__confirmed_boms = set([])
+    self.__prune_boms = []
+    self.__prune_jars = {}
+    self.__prune_debians = {}
+    self.__prune_containers = {}
+    self.__prune_gce_images = {}
+    self.__invalid_versions = {}
+
+  def audit_artifacts(self):
+    self.audit_bom_services(self.__released_boms, 'released')
+    self.audit_bom_services(self.__unreleased_boms, 'unreleased')
+    self.audit_package(
+        'jar', self.__jar_versions, self.__unused_jars)
+    self.audit_package(
+        'debian', self.__debian_versions, self.__unused_debians)
+    self.audit_package(
+        'container', self.__container_versions, self.__unused_containers)
+    self.audit_package(
+        'image',
+        self.__gce_image_versions, self.__unused_gce_images)
+
+    def maybe_write_log(what, data):
+      if not data:
+        return
+      path = os.path.join(self.get_output_dir(), 'audit_' + what + '.yml')
+      logging.info('Writing %s', path)
+      write_to_path(
+          yaml.dump(data, allow_unicode=True, default_flow_style=False),
+          path)
+
+    confirmed_boms = self.__all_bom_versions - set(self.__invalid_boms.keys())
+    maybe_write_log('missing_debians', self.__missing_debians)
+    maybe_write_log('missing_jars', self.__missing_jars)
+    maybe_write_log('missing_containers', self.__missing_containers)
+    maybe_write_log('missing_images', self.__missing_images)
+    maybe_write_log('found_debians', self.__found_debians)
+    maybe_write_log('found_jars', self.__found_jars)
+    maybe_write_log('found_containers', self.__found_containers)
+    maybe_write_log('found_images', self.__found_images)
+    maybe_write_log('unused_debians', self.__unused_debians)
+    maybe_write_log('unused_jars', self.__unused_jars)
+    maybe_write_log('unused_containers', self.__unused_containers)
+    maybe_write_log('unused_images', self.__unused_gce_images)
+    maybe_write_log('invalid_boms', self.__invalid_boms)
+    maybe_write_log('confirmed_boms', sorted(list(confirmed_boms)))
+    maybe_write_log('invalid_versions', self.__invalid_versions)
+
+  def most_recent_version(self, name, versions):
+    """Find the most recent version built."""
+    if not versions:
+      return None
+    raw_versions = set([version.split('-')[0] for version in versions])
+    sem_vers = []
+    for text in raw_versions:
+      try:
+        sem_vers.append(SemanticVersion.make('version-' + text))
+      except Exception as ex:
+        bad_list = self.__invalid_versions.get(name, [])
+        bad_list.append(text)
+        self.__invalid_versions[name] = bad_list
+        logging.error('Ignoring invalid %s version "%s": %s', name, text, ex)
+    return sorted(sem_vers, cmp=SemanticVersion.compare)[-1].to_version()
+
+  def determine_prunings(self):
+    def filter_not_version(version, version_list):
+      return [value for value in version_list
+              if not value.startswith(version)
+              or value < version + '-20180201']
```

Comment:
Datestring should be parameterized here, right?

Reply:
Yea, but not yet. Actually this wasnt supposed to survive. It was more explicit. There are some "18" boms during a transition period, so < 2018 is going to prune those. None of those were released to my knowledge so we can prune them out after a bit more time and add the option.